### DISCUSSION TITLE
Particle Life (C++) New Release

### DISCUSSION THREAD

**Discussion by hunar4321** [hunar4321/particle-life#80] [Announcements]
New features by @ker2x:
- Major GUI changes & improvements
- Performance improvements
- Experimental features like gradual mutation of parameters

![cover](https://user-images.githubusercontent.com/39065447/202216128-c4eb2a7d-ed9d-4978-ba4b-7e5dd166f2fb.jpg)

# Download binaries: 
1. For mac OS check out: https://github.com/ker2x/particle-life/releases/tag/osx-v1
2. Windows binary is here: 

<hr /><em>This discussion was created from the release <a href='https://github.com/hunar4321/particle-life/releases/tag/v.0.3.0'>Particle Life</a>.</em>

**Comment by hunar4321** [comment]
Note that some of the patterns shown in the original video might not reproduce exactly with this release but now the randomize button can produce interesting patterns most of the time even without too much fine-tuning!

Another improvement can be added to the future releases is to add the ability to produce more color types just like the online JavaScript version [here ](https://hunar4321.github.io/particle-life/particle_life.html#91651088029)  but that will increase the number of interactions exponentially. In this case, one can only rely on randomize button to produce patterns instead of manual fine-tuning.
In my opinion, careful manual fine-tuning can still produce way more interesting patterns than brute force search through randomize button.

**Comment by Myrkiriad-coder** [comment]
There seems to be missing c++ files. Where are ofMain.h" and "ofxGui.h" ?

**Reply by hunar4321** [reply]
To compile the C++ project, you need to add the OpenFrameworks library to the project. The setup procedure for OF is described in the readme here: https://github.com/hunar4321/particle-life/blob/main/README.md#code

**Reply by Myrkiriad-coder** [reply]
Thanks.

I managed to compile and run the C++ version. However it runs only at 5 FPS on my PC intel(R) Core(TM) i5-7200U CPU @ 2.50GHz.

The html runs much faster at 25 FPS on the same PC.

Any idea why ?

**Reply by ker2x** [reply]
compile it in release mode, enable openMP, enable fast-math, enable avx2

**Reply by ker2x** [reply]
with various compilation flag, I run it at 100FPS on my laptop

**Reply by Myrkiriad-coder** [reply]
Thanks now it runs fast.

**Reply by KhadrasWellun** [reply]
> compile it in release mode, enable openMP, enable fast-math, enable avx2

Where can I find this option to enable them?

**Reply by ker2x** [reply]
in visual studio :

- right click on particle_life (the project, not the solution)
- property (at the bottom)
- C/C++
- language : openmp support : Yes
- code generation : floating point model : fast
- code generation : enable instruction set : (your instruction set, for your cpu you can enable AVX2)
- optimization : maximum optimization (/O2)
- optimization : whole program optimization

then "rebuild all" (you don't have to rebuild all every time, but since you just changed a lot of stuff you should do it at least once)

**Reply by KhadrasWellun** [reply]
Thanks! 
![Interface 1 2](https://user-images.githubusercontent.com/116345441/205457306-1ec851a6-3868-4f05-b61c-552c05dc3e46.png)
I have 56 FPS with 8432 particles of 8 colors.

**Reply by ker2x** [reply]
ho you added more color ? :)

**Reply by KhadrasWellun** [reply]
> ho you added more color ? :)
I added 4 more colors, and I added more minor changes like: random viscosity, random particle numbers with different buton, random probability, viscosity and probability evolution, I reordered some menu options. 
More to come: 
I will add for each type of interaction both an attraction and a repulsion matrix that will act simultaneously on the same interaction, and each will have a radius matrix and a probability matrix. 
I will also add metaball, gravity, electromagnetic force acting only on particle structures (metaballs), dark matter, dark energy, quantum bonds between particles, etc.
Now I've added dark particles, which are just the color of the background. The aim was that they could not be seen but would act on the other lighter coloured particles. It's a kind of simulation of dark matter, but it's not what I want it to be in the end.
My modifications are here:
https://github.com/KhadrasWellun/particle-life

**Reply by hunar4321** [reply]
> > ho you added more color ? :)
> > I added 4 more colors, and I added more minor changes like: random viscosity, random particle numbers with different buton, random probability, viscosity and probability evolution, I reordered some menu options.
> > More to come:
> > I will add for each type of interaction both an attraction and a repulsion matrix that will act simultaneously on the same interaction, and each will have a radius matrix and a probability matrix.
> > I will also add metaball, gravity, electromagnetic force acting only on particle structures (metaballs), dark matter, dark energy, quantum bonds between particles, etc.
> > Now I've added dark particles, which are just the color of the background. The aim was that they could not be seen but would act on the other lighter coloured particles. It's a kind of simulation of dark matter, but it's not what I want it to be in the end.
> > My modifications are here:
> > https://github.com/KhadrasWellun/particle-life

@KhadrasWellun I had a look at your video demo: https://youtu.be/b-afHvgact0
This is impressive! Did the unstable cells just start dividing at the end of the video without you changing any parameters? If so, this is really a nice model for cell reproduction

**Reply by KhadrasWellun** [reply]
Yes, the cell begin to diveded without any modification of the parameters. It start to move chaotic and then duplicates.

**Comment by ker2x** [comment]
I bought a new laptop (Asus ZenBook EVO i5). 130 to 160FPS on the "dna" interesting model :)

**Reply by Myrkiriad-coder** [reply]
What is the "dna" model ?

**Reply by ker2x** [reply]
it's one of the model included in the "interesting model" you can load in the C++ application
